eth_p2p/rlpx_protocols/shh_protocol.nim

```diff
@@ -416,31 +539,406 @@ proc add*(self: var Queue, msg: Message) =
+      return
 
-  proc powRequirement(peer: Peer, value: float64) =
-    discard
+    for envelope in envelopes:
+      # check if expired or in future, or ttl not 0
+      if not envelope.valid():
+        warn "Expired or future timed envelope"
+        # disconnect from peers sending bad envelopes
+        # await peer.disconnect(SubprotocolReason)
+        continue
+
+      let msg = initMessage(envelope)
+      if not msg.allowed(peer.networkState.config):
+        # disconnect from peers sending bad envelopes
+        # await peer.disconnect(SubprotocolReason)
+        continue
+
+      # This peer send it thus should not receive it again
+      peer.state(shh).received.incl(msg)
+
+      if peer.networkState.queue.add(msg):
+        # notify filters of this message
+        peer.networkState.filters.notify(msg)
+
+  proc powRequirement(peer: Peer, value: uint) =
+    if not peer.state.initialized:
+      warn "Handshake not completed yet, discarding powRequirement"
+      return
+
+    peer.state.powRequirement = cast[float64](value)
 
   proc bloomFilterExchange(peer: Peer, bloom: Bytes) =
-    discard
+    if not peer.state.initialized:
+      warn "Handshake not completed yet, discarding bloomFilterExchange"
+      return
+
+    peer.state.bloom.bytesCopy(bloom)
 
   nextID 126
 
   proc p2pRequest(peer: Peer, envelope: Envelope) =
+    # TODO: here we would have to allow to insert some specific implementation
+    # such as e.g. Whisper Mail Server
     discard
 
   proc p2pMessage(peer: Peer, envelope: Envelope) =
-    discard
+    if peer.state.trusted:
+      # when trusted we can bypass any checks on envelope
+      let msg = Message(env: envelope, isP2P: true)
+      peer.networkState.filters.notify(msg)
+
+# 'Runner' calls ---------------------------------------------------------------
+
+proc processQueue(peer: Peer) =
+  var envelopes: seq[Envelope] = @[]
+  for message in peer.networkState(shh).queue.items:
+    if peer.state(shh).received.contains(message):
+      # debug "message was already send to peer"
+      continue
+
+    if message.pow < peer.state(shh).powRequirement:
+      debug "Message PoW too low for peer"
+      continue
+
+    if not bloomFilterMatch(peer.state(shh).bloom, message.bloom):
+      debug "Message does not match peer bloom filter"
+      continue
+
+    debug "Adding envelope"
+    envelopes.add(message.env)
+    peer.state(shh).received.incl(message)
+
+  debug "Sending envelopes", amount=envelopes.len
+  # await peer.messages(envelopes)
+  asyncCheck peer.messages(envelopes)
+
+proc run(peer: Peer) {.async.} =
+  peer.state(shh).running = true
+  while peer.state(shh).running:
+    if not peer.networkState(shh).config.isLightNode:
+      peer.processQueue()
+    await sleepAsync(300)
+
+proc pruneReceived(node: EthereumNode) =
+  if node.peerPool != nil: # XXX: a bit dirty to need to check for this here ...
+    for peer in node.peers(shh):
+      if not peer.state(shh).initialized:
```

Comment:
I usually prefer to cache the `state` and `networkState` calls to variables, even if they are not particularly expensive.

Reply:
OK, I will create vars for all the states